jooby-filewatcher/src/main/java/org/jooby/filewatcher/FileWatcher.java

```diff
@@ -0,0 +1,385 @@
+ * <p>
+ * You can specify a {@link FileEventHandler} instance or class while registering a path:
+ * </p>
+ *
+ * <p>
+ * Instance:
+ * </p>
+ * <pre>{@code
+ * {
+ *   use(new FileWatcher()
+ *     .register(Paths.get("mydir"), (kind, path) -> {
+ *       log.info("found {} on {}", kind, path);
+ *     });
+ *   );
+ * }
+ * }</pre>
+ *
+ * <p>
+ * Class reference:
+ * </p>
+ *
+ * <pre>{@code
+ * {
+ *   use(new FileWatcher()
+ *     .register(Paths.get("mydir"), MyFileEventHandler.class)
+ *   );
+ * }
+ * }</pre>
+ *
+ * <p>
+ * Worth to mention that <code>MyFileEventHandler</code> will be provided by Guice.
+ * </p>
+ *
+ * <h2>options</h2>
+ * <p>
+ * You can specify a couple of options at registration time:
+ * </p>
+ * <pre>{@code
+ * {
+ *   use(new FileWatcher(
+ *     .register(Paths.get("mydir"), MyFileEventHandler.class, options -> {
+ *       options.kind(StandardWatchEventKinds.ENTRY_MODIFY)
+ *              .recursive(false)
+ *              .includes("*.java");
+ *     });
+ *   ));
+ * }
+ * }</pre>
+ *
+ * <p>
+ * 1. Here we listen for {@link StandardWatchEventKinds#ENTRY_MODIFY} (we don't care about create or
+ * delete).
+ * </p>
+ * <p>
+ * 2. We turn off recursive watching, only direct files are detected.
+ * </p>
+ * <p>
+ * 3. We want <code>.java</code> files and we ignore any other files.
+ * </p>
+ *
+ * <h2>configuration file</h2>
+ * <p>
+ * In addition to do it programmatically, you can do it via configuration properties:
+ * </p>
+ *
+ * <pre>{@code
+ * {
+ *   use(new FileWatcher()
+ *     .register("mydirproperty", MyEventHandler.class)
+ *   );
+ * }
+ * }</pre>
+ *
+ * <p>
+ * The <code>mydirproperty</code> property must be present in your <code>.conf</code> file.
+ * </p>
+ *
+ * <p>
+ * But of course, you can entirely register paths from <code>.conf</code> file:
+ * </p>
+ *
+ * <pre>{@code
+ * filewatcher {
+ *   register {
+ *     path: "mydir"
+ *     handler: "org.example.MyFileEventHandler"
+ *     kind: "ENTRY_MODIFY"
+ *     includes: "*.java"
+ *     recursive: false
+ *   }
+ * }
+ * }</pre>
+ *
+ * <p>
+ * Multiple paths are supported using array notation:
+ * </p>
+ *
+ * <pre>{@code
+ * filewatcher {
+ *   register: [{
+ *     path: "mydir1"
+ *     handler: "org.example.MyFileEventHandler"
+ *   }, {
+ *     path: "mydir2"
+ *     handler: "org.example.MyFileEventHandler"
+ *   }]
+ * }
+ * }</pre>
+ *
+ * <p>
+ * Now use the module:
+ * </p>
+ * <pre>{@code
+ * {
+ *   use(new FileWatcher());
+ * }
+ * }</pre>
+ *
+ * <p>
+ * The {@link FileWatcher} module read the <code>filewatcher.register</code> property and setup
+ * everything.
+ * </p>
+ *
+ * @author edgar
+ */
+public class FileWatcher implements Module {
+  private static final Consumer<FileEventOptions> EMPTY = it -> {
+  };
+
+  /** The logging system. */
+  private final Logger log = LoggerFactory.getLogger(getClass());
+
+  private List<CheckedFunction2<Config, Binder, FileEventOptions>> bindings = new ArrayList<>();
```

Comment:
In FileMonitor keys and options fields are both final and this field is very similar to them, do you have a rule when final should be used for fields?